### PR TITLE
Spark FileSystems should be closed when no longer needed.

### PR DESCRIPTION
**Pull request by tsmock**
This fixes a resource leak, which will improve resource usage.

### Description:

Put `org.apache.hadoop.fs.FileSystem` creation into `try-with-resource` blocks. This ensures that the `FileSystem` is properly closed. There are also some fixes for sonar sprinkled in (mostly deduplication of strings).

### Potential Impact:

Major impact: Less resource utilization
No impact: No new exceptions are thrown (those that are thrown through the try-with-resources are caught and handled silently, to avoid breaking downstream users).

### Unit Test Approach:

N/A

### Test Results:

With https://github.com/osmlab/atlas-checks/pull/398, I was able to reduce the memory required to run the test check on NZL from 15 GB to 1GB. It may have been possible to run it with less memory (so < 1GB).

Notes:
* Additional modifications to `atlas-checks` was required in order to run the current `dev` branch of `atlas-generator`. Specifically, the resolution strategy for `jackson` had to be commented out (due to 54d1bd2b607f430379e18e2fee9166f6d803bf94 ).

------

In doubt: [Contributing Guidelines](https://github.com/osmlab/atlas/blob/dev/CONTRIBUTING.md)

TODO list:
List of everything that uses `org.apache.hadoop.fs.FileSystem` (from imports):

* ~src/main/java/org/openstreetmap/atlas/generator/PbfLocator.java~ [DONE, implements AutoCloseable]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/AbstractFileOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/MultipleAtlasCountryStatisticsOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/MultipleAtlasFeatureChangeOutput.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/MultipleAtlasGeoJsonOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/MultipleAtlasOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/MultipleAtlasProtoOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/MultipleAtlasStatisticsOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/MultipleFeatureChangeOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/MultipleLineDelimitedGeojsonOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/delta/AddedMultipleAtlasDeltaOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/delta/ChangedMultipleAtlasDeltaOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/delta/MultipleAtlasDeltaOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/persistence/delta/RemovedMultipleAtlasDeltaOutputFormat.java~ [N/A -- FileSystem passed into class]
* ~src/main/java/org/openstreetmap/atlas/generator/tools/filesystem/FileSystemCreator.java~ [N/A -- creates FileSystem objects for return]
* ~src/main/java/org/openstreetmap/atlas/generator/tools/filesystem/FileSystemHelper.java~ [DONE]
* ~src/main/java/org/openstreetmap/atlas/generator/tools/spark/DataLocator.java~ [DONE]
* ~src/main/java/org/openstreetmap/atlas/generator/tools/spark/SparkJob.java~ [DONE]
* ~src/main/java/org/openstreetmap/atlas/generator/tools/spark/input/SparkInput.java~ [DONE]
* ~src/main/java/org/openstreetmap/atlas/generator/tools/spark/rdd/ShardedAtlasRDDLoader.java~ [DONE]
* ~src/main/java/org/openstreetmap/atlas/generator/tools/spark/utilities/SparkFileHelper.java~ [DONE]
* src/main/java/org/openstreetmap/atlas/generator/tools/streaming/ResourceFileSystem.java
* ~src/main/java/org/openstreetmap/atlas/generator/tools/streaming/resource/HDFSFile.java~ [DONE, implements AutoCloseable]
* ~src/main/java/org/openstreetmap/atlas/generator/tools/streaming/resource/HDFSWalker.java~ [mostly DONE, implements AutoCloseable, but I need to decide what to do with the exception, throw or catch, log, and suppress]

### CODE REVIEW

**Comment by atiannicelli** [comment]
Nice find @tsmock !!

**Comment by tsmock** [comment]
> Nice find @tsmock !!

I'm going through the rest of the files that use `org.apache.hadoop.fs.FileSystem` right now. `FileSystemHelper` just happened to be the major user of memory in a heap dump.

**Review by matthieun — COMMENTED** [review]
This is a great find, thanks @tsmock!
I'll waif to see the other classes you find that need this improvement.

File: src/main/java/org/openstreetmap/atlas/generator/tools/filesystem/FileSystemHelper.java [hunk 1]
```diff
@@ -428,7 +447,7 @@ public static WritableResource writableResource(final String path,
         }
         catch (final Exception e)
         {
-            throw new CoreException("Unable to read {}", hadoopPath, e);
+            throw new CoreException(UNABLE_TO_READ, hadoopPath, e);
```
Comment (adahn6) [inline src/main/java/org/openstreetmap/atlas/generator/tools/filesystem/FileSystemHelper.java:450]: 💯thanks for cleaning up the Strings here!

**Comment by tsmock** [comment]
> I'll waif to see the other classes you find that need this improvement.

I don't know about "need", but so far `SparkFileHelper` has the same problem, `HDFSWalker`/`HDFSIterator` (for these, I'm implementing `AutoCloseable`, since they may be stored and reused). I'll need to do another run looking for anything I implement `AutoCloseable` for, but step-by-step.

File: src/main/java/org/openstreetmap/atlas/generator/tools/filesystem/FileSystemHelper.java (hunk 1 above)
Comment (tsmock) [reply src/main/java/org/openstreetmap/atlas/generator/tools/filesystem/FileSystemHelper.java:450]: Eclipse Sonar Lint was complaining about it, and I know that `atlas-checks` uses sonar, so I figured I'd save a round trip on that. Hopefully.

**Review by lucaspcram — APPROVED** [review]
This is great!

**Comment by tsmock** [comment]
> This is great!

Sorry about dismissing your review -- I was originally intending on just modifying the `FileSystemHelper` class, so I didn't create this as a draft PR. In retrospect, I should have done that.

**Comment by tsmock** [comment]
I'm now working on test code, so the "main" portion of the code is ready for review.

Specifically, how would the `atlas-generator` project like to handle https://github.com/osmlab/atlas-generator/pull/167/files#diff-d71e9d8dbdf395178a6c991f67801a122152eefe2ebaeba5261be57c5a8ce665R216 (comment is `TODO add a logger or throw?`)?

Up to this point, `HDFSWalker` has not had a logger. In this case, I'm inclined to add a logger instead of rethrowing the exception for downstream users.

**Comment by tsmock** [comment]
I haven't found a reliable way to check for memory leaks, so I cannot write tests for that. :(

Anyway, I've added a logger to HDFSWalker to avoid throwing a new (unexpected) exception.

I think I'm done with this PR. I just ran a PMD analysis on atlas-generator, and the following resources aren't closed:
| File | Line |
|-----|-----|
| src/main/java/org/openstreetmap/atlas/generator/tools/filesystem/FileSystemHelper.java | 372 |
| src/main/java/org/openstreetmap/atlas/generator/tools/spark/input/SparkInput.java | 319 |

Both of the above issues are (kind of) false positives. The remaining one in FileSystemHelper is in a `WritableResource`, which probably should implement AutoCloseable and close the `OutputStream` on L372. Fortunately, I'll be able to implement the interface without breaking downstream source compatibility in Atlas.

For `SparkInput`, the calling function should deal with that, since it is a `FileSystem` inside an `Optional`. The sole user of that method is `transform` inside the same class, which does get the `fileSystem` inside a `try-with-resources` block.

**Comment by vladlemberg** [comment]
@matthieun, is it possible to release Java8 version with this fix ?

**Comment by atiannicelli** [comment]
@lucaspcram @matthieun @adahn6  Do you have a timeframe on when this will be reviewed? There are a couple atlas checks PRs that depend on this fix. Thanks!

**Review by lucaspcram — APPROVED** [review]
Gave it one more look, thanks for the additions! And for the reminder that this was still open ;)

**Pull request merged**